### PR TITLE
Reduced flicker on pdf reload and slightly improved painting performance

### PR DESCRIPTION
**Pull request by a-reisberg**
See #572

### CODE REVIEW

**Comment by sunderme** [comment]
I still think that

'#if QT_VERSION >= 0x050000
￼	        else {
￼	            QTimer::singleShot(delayTimeout, obj, [=]() {QMetaObject::invokeMethod(info.obj, info.slot, Q_ARG(QPixmap, img), Q_ARG(int, pageNr));});
￼	        }
￼	#endif
￼	    }'

does not do what it is supposed to do. Please take it out.

**Comment by a-reisberg** [comment]
Could you please elaborate? I tested (note the `pgrNr + 100` below)
```C++
if (delayTimeout >= 0) {
    if (!img.isNull())
        QMetaObject::invokeMethod(info.obj, info.slot, Q_ARG(QPixmap, img), Q_ARG(int, pageNr + 100));
#if QT_VERSION >= 0x050000
    else {
        QTimer::singleShot(delayTimeout, obj, [=]() {QMetaObject::invokeMethod(info.obj, info.slot, Q_ARG(QPixmap, img), Q_ARG(int, pageNr));});
    }
#endif
}
and then `qDebug() << pageNr` inside `PdfWidget::setImage` and got the correct value.
```

**Comment by benibela** [comment]
You can just create a new slot (any method can be marked as slot when the class is derived from QObject ), then it also works on qt4

And use tabs instead of spaces

Once it does not flicker anymore, it would also be easier to implement live previewing (e.g. for #550)

**Comment by sunderme** [comment]
how is it supposed to work ?
you call setImage after delayTimeout, with no image yet calculated. There is another singleShot timer which call update which draws the background if the image is not present. So I don't see what this call is supposed to do and the syntax is odd, even considering the new Qt5 syntax (inokeMethod is calling a slot but you already are using a method (singleShot) which can directly address slots, so it is the same thing done twice)
Even it might work for now, no guarantees that it works on all qt versions properly.

Please, just remove that part of the code.

**Comment by a-reisberg** [comment]
If you look at how `setImage` is written:
```C++
void PDFWidget::setImage(QPixmap, int pageNr)
{
    qDebug() << pageNr;
	forceUpdate = true;
    update(pageRect(pageNr));
}
```
it doesn't even use the `img` argument (in fact, before it didn't even use the `pageNr` argument). It's just used as a signaling device all along. So in the case where `img` hasn't been computed yet, `setImage` will be called with a null `img`, which is fine, since it doesn't use it at all. `setImage` will then call `update()`, which will call `renderToImage` again to actually render that page.

Now, that's the path for qt5. For qt4, those lines won't even be compiled. So for `qt4` and below, a less refined path is used inside `delayedUpdate()` itself, by having a timer which calls `update()`:
```C++
// Update eventually if loading takes too long. For Qt5 and above, lambda expression for SLOT has been used to handle
// this within renderToImage itself.
#if (QT_VERSION < 0x050000)
    QTimer::singleShot(1000, this, SLOT(update()));
#endif
```

Now, according to @benibela, it can also be made to work for qt4. @benibela Could you please elaborate? I don't know much about Qt. Thanks!

**Comment by sunderme** [comment]
the problem with a slot is that you can't send parameters from singleshot without the qt5/lambda methodology ...

**Comment by a-reisberg** [comment]
@sunderme I just had an idea of how to do it. Let me try...

**Comment by sunderme** [comment]
by the way, the singleshot(...,functor) has only been introduced since qt 5.4 ...

**Comment by a-reisberg** [comment]
Ok, now it should work with all Qt versions.

File: src/pdfviewer/pdfrendermanager.cpp [hunk 1]
```diff
@@ -25,6 +25,20 @@ const int kMaxPageZoom = 1000000;
 const qreal kMaxDpiForFullPage = 1000.0;
 
 
+SetImageForwarder::SetImageForwarder(QObject *parent, QObject *obj, const char *rec, QPixmap img, int pageNr):
+    QObject(parent), obj(obj), rec(rec), img(img), pageNr(pageNr)
+{
+}
+
+void SetImageForwarder::setImage() {
+    QMetaObject::invokeMethod(obj, rec, Q_ARG(QPixmap, img), Q_ARG(int, pageNr + 100));
```
Comment (sunderme) [inline src/pdfviewer/pdfrendermanager.cpp]: where is the pageNr+100 coming from ?

Comment (a-reisberg) [reply src/pdfviewer/pdfrendermanager.cpp]: Ooops. Sorry. That's from debugging session. Will fix now.

**Pull request merged**